src/lib.rs

```diff
@@ -11,6 +11,1388 @@
+                    format!("Unable to subscribe to topic: {e:?}"),
+                )
+            })?;
+
+        Ok(())
+    }
+
+    /// Unsubscribe the mqtt client to a topic.
+    ///
+    /// # Arguments
+    /// * `topic` - Topic to subscribe to.
+    async fn unsubscribe(&self, topic: &str) -> Result<(), UStatus> {
+        self.inner_mqtt_client
+            .unsubscribe(topic)
+            .await
+            .map_err(|e| {
+                UStatus::fail_with_code(
+                    UCode::INTERNAL,
+                    format!("Unable to unsubscribe from topic: {e:?}"),
+                )
+            })?;
+
+        Ok(())
+    }
+}
+
+/// Configuration for the mqtt client.
+pub struct MqttConfig {
+    /// Port of the mqtt broker to connect to.
+    pub mqtt_port: String,
+    /// Hostname of the mqtt broker.
+    pub mqtt_hostname: String,
+    /// Max buffered messages for the mqtt client.
+    pub max_buffered_messages: i32,
+    /// Max subscriptions for the mqtt client.
+    pub max_subscriptions: i32,
+    /// Session Expiry Interval for the mqtt client.
+    pub session_expiry_interval: i32,
+    /// Optional SSL options for the mqtt connection.
+    pub ssl_options: Option<mqtt::SslOptions>,
+}
+
+/// UP Client for mqtt.
+pub struct UPClientMqtt {
+    /// Client instance for connecting to mqtt broker.
+    mqtt_client: Arc<dyn MockableMqttClient>,
+    /// Map of subscription identifiers to subscribed topics.
+    subscription_topic_map: Arc<RwLock<HashMap<i32, String>>>,
+    /// Map of topics to listeners.
+    topic_listener_map: Arc<RwLock<HashMap<String, HashSet<ComparableListener>>>>,
+    /// My authority
+    authority_name: String,
+    /// Indicates where client instance is running.
+    client_type: UPClientMqttType,
+    /// List of free subscription identifiers to use for the client subscriptions.
+    free_subscription_ids: Arc<RwLock<HashSet<i32>>>,
+}
+
+/// Type of UPClientMqtt.
+pub enum UPClientMqttType {
+    Device,
+    Cloud,
+}
+
+impl UPClientMqtt {
+    /// Create a new UPClientMqtt.
+    ///
+    /// # Arguments
+    /// * `config` - Configuration for the mqtt client.
+    /// * `client_id` - Client id for the mqtt client.
+    /// * `authority_name` - Authority name for the mqtt client.
+    /// * `client_type` - Type of client instance.
+    pub async fn new(
+        config: MqttConfig,
+        client_id: UUID,
+        authority_name: String,
+        client_type: UPClientMqttType,
+    ) -> Result<Self, UStatus> {
+        let subscription_topic_map = Arc::new(RwLock::new(HashMap::new()));
+        let topic_listener_map = Arc::new(RwLock::new(HashMap::new()));
+        let free_subscription_ids =
+            Arc::new(RwLock::new((1..config.max_subscriptions + 1).collect()));
+
+        let subscription_topic_map_handle = subscription_topic_map.clone();
+        let topic_listener_map_handle = topic_listener_map.clone();
+
+        let mqtt_client = AsyncMqttClient::new_client(
```

Comment:
Have you considered passing in a `Arc<dyn MockableMqttClient>` in order to make this possible to be injected by the client / user of the library?